### PR TITLE
[bugfix]: stabilize GenRL reward and PPO training

### PR DESCRIPTION
**Pull request by Abecid**
## Purpose

Stabilize the GenRL / diffusion RL training path on top of `py/add_rl`.

This PR focuses on getting the RL pipeline to run more reliably for Wan T2V training with multi-reward scoring, especially HPSv3 + VideoAlign rewards, and reducing PPO/GRPO instability from noisy rollout grouping and reward-model memory pressure.

Fixes #

## Changes

- Keep repeated samples from the same prompt together in `DistributedKRepeatSampler`.
  - Each rank now receives whole prompt groups when `sample_batch_size` is divisible by `num_video_per_prompt`.
  - This avoids splitting one GRPO comparison group across ranks with different sampled SDE timestep windows.

- Add safer PPO/GRPO training behavior.
  - Factor PPO loss and diagnostics into a helper.
  - Add optional rollout microbatch accumulation via `accumulate_ppo_microbatches`.
  - Add post-update logprob/KL diagnostics.
  - Add `flash_tgr` loss reweighting mode to avoid overly large LongCat timestep reweighting during full-parameter training.
  - Lower the LongCat config clip range to `1e-4`.

- Improve reward stack robustness.
  - Move decoded rollout videos to CPU before reward scoring/logging.
  - Avoid asynchronous GPU reward scoring when reward models are on GPU, preventing VideoAlign from overlapping with Wan denoising and causing CUDA OOM.
  - Clear cached GPU reward inferencers before PPO training.
  - Patch VideoAlign/HPSv3 compatibility issues for current Qwen2-VL / transformers behavior.

- Add/adjust GenRL training configs.
  - Keep full fine-tuning as the default path for the Wan LongCat-style multi-reward config.
  - Keep LoRA options configurable.
  - Add fixed eval reward settings for more stable reward tracking.

## Running Experiments

Example local/multi-GPU launch:

```bash
bash examples/train/run.sh \
  examples/train/configs/genrl_wan2.1_t2v_1.3B_longcat.yaml \
  --method.prompt_dataset_path /path/to/filtered_prompts \
  --training.checkpoint.output_dir /path/to/outputs/genrl_longcat
```

## Experiment Environment Notes

I verified the GenRL run in a Modal environment roughly equivalent to:

- Base image: `nvidia/cuda:12.8.1-devel-ubuntu22.04`
- Python: `3.12`
- GPU stack: CUDA 12.8, PyTorch/torchvision/torchaudio from the cu128 index
- System packages: `ffmpeg`, `libgl1`, `libglib2.0-0`, `build-essential`, `ninja-build`, `cmake`, `git-lfs`
- Extra Python deps used by the reward stack:
  - `datasets==3.6.0`
  - `matplotlib==3.10.3`
  - `peft==0.10.0`
  - `qwen-vl-utils==0.0.11`
  - `safetensors==0.5.3`
  - `timm==1.0.15`
  - `trl==0.8.6`
  - `numpy==1.26.4`
  - `scipy==1.15.2`
- FlashAttention: `flash_attn-2.8.3+cu128torch2.10` prebuilt wheel

The Modal launcher itself is not included in this PR; these notes are only to document the environment used for testing/debugging.

## Performance

Confirmed the following reward metrics going up over 40 iterations:

- reward_videoalign_ta_raw: 1.5
- reward_videoalign_mq_raw: 0.6
- reward_hpsv3_percentile_raw: 8.67
- reward_hpsv3_general_raw: 3.456

### CODE REVIEW

File: fastvideo/train/methods/rl/genrl.py [hunk 1]
```diff
@@ -135,6 +140,14 @@ def __init__(
...
         self._rank = wg.rank
         self._is_main = wg.rank == 0
+        total_samples = self._world_size * self._sample_batch_size
+        if total_samples % self._num_video_per_prompt != 0:
+            raise ValueError(
+                "world_size * sample_batch_size must be divisible by "
+                "num_video_per_prompt for DistributedKRepeatSampler. Got "
+                f"{self._world_size} * {self._sample_batch_size} and "
+                f"{self._num_video_per_prompt}."
+            )
```
[the remark this answers is not in the document]
Comment (Davids048) [reply fastvideo/train/methods/rl/genrl.py:150]: this is tied to the first change. Keep repeated samples from the same prompt together in DistributedKRepeatSampler.
should patch to enforce the rule

File: fastvideo/train/methods/rl/genrl.py [hunk 2]
```diff
@@ -831,6 +1007,107 @@ def _ppo_train(
             info: dict[str, list] = defaultdict(list)
             _ppo_batch_times: list[float] = []
 
+            if (
+                self._accumulate_ppo_microbatches
+                and len(batched_list) > 1
+            ):
+                probe_args = None
+                for j in self._train_timesteps:
```
[the remark this answers is not in the document]
Comment (Davids048) [reply fastvideo/train/methods/rl/genrl.py:1015]: robustness

**Review by Davids048 — CHANGES_REQUESTED** [review]
Currently the PR combines runtime/OOM fixes, reward-model compatibility patches, sampler behavior, PPO/GRPO training-loop changes, LoRA support, and config/default changes. That makes it hard to attribute regressions in the future.

**I would recommend splitting this into a small stack of PRs. For example:**
- OOM fixes
- Reward model compatibility fixes
- prompt group placement
- training loop changes
- LoRA support for GenRL
(Don't need to follow this exactly tho.)
There is an agent skill that may help: [link](https://github.com/hao-ai-lab/FastVideo/blob/main/.agents/skills/decompose-pipeline-pr/SKILL.md?plain=1)



**The Gemini comments look valid to me and should be addressed before merge.** 

3 additional issues I think are important:

## 1. LoRA is inserted after the Wan transformer has already gone through the FSDP/HSDP loader path

In `fastvideo/train/models/wan/wan_genrl.py`, `GenRLWanModel.__init__()` calls the parent `WanModel` init first, which loads the transformer through `load_module_from_path()`. After that, when `use_lora=True`, `_apply_fastvideo_lora()` replaces modules and creates new LoRA parameters.

That means the LoRA parameters are introduced **after** the transformer has already gone through the distributed loader/sharding path. Those newly inserted parameters may not be managed with the same FSDP/HSDP device mesh or checkpointing assumptions as the rest of the model. 

## 2. Synchronous GPU reward mode still retains decoded CPU videos for all rollout batches before scoring

With `reward_on_gpu: true`, async reward scoring is disabled, but `sample_epoch()` still stores decoded `videos_cpu` for each rollout batch and delays reward scoring until after all batches are sampled. With `num_batches_per_epoch: 4`, this can retain multiple large decoded video tensors on CPU at once.


## 3. Config naming

`examples/train/configs/genrl_wan2.1_t2v_1.3B_longcat.yaml` is confusing because FastVideo already has an actual LongCat-Video model. The "longcat" should be more specific and point to the reward method(s) used.Please rename it to something more explicit. 

It is fine to document inside the file that the recipe is derived from GenRL's LongCat config. 

## Test
We should add testing (CI/unit tests) on the roadmap for this development.

**Comment by Abecid** [comment]
Thanks for the review. I split this mega PR into six focused PRs so each concern can be reviewed independently:

1. #1400 — reward model compatibility
2. #1401 — prompt group placement / sampler behavior
3. #1402 — runtime and memory stability helpers
4. #1403 — PPO/GRPO training-loop fixes and Gemini review comments
5. #1404 — GenRL Wan LoRA adapter support
6. #1405 — explicit HPSv3 + VideoAlign config recipes / config rename

I also addressed the Gemini comments in the split PRs:
- sampler validation / prompt grouping: #1401
- accumulated microbatch `zero_grad`, per-timestep probe args, and post-update KL diagnostics: #1403
- LoRA missing-weight runtime guard: #1404

For your three additional issues:
- LoRA/FSDP insertion order is isolated in #1404 for focused review.
- GPU reward memory behavior is isolated in #1402.
- The confusing `longcat` config naming is handled in #1405.

I’m going to treat #1391 as superseded by this stack unless you prefer I keep it open for reference.

**Comment by SolitaryThinker** [comment]
closing for now, feel free to repo
